From fec0ff571f91b489fb9ef07cb493ba1b03563371 Mon Sep 17 00:00:00 2001
From: Doug Ledford <dledford@redhat.com>
Date: Wed, 24 Sep 2014 15:22:19 -0400
Subject: [infiniband] ocrdma: get vlan tag from ib_qp_attrs

Message-id: <3a3211e1c2ea1fe60f66a5785b9ae3adfd001d3e.1411571481.git.dledford@redhat.com>
Patchwork-id: 96649
O-Subject: [Patch RHEL7 44/48] RDMA/ocrdma: get vlan tag from ib_qp_attrs
Bugzilla: 1081067 1088577
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Selvin Xavier <selvin.xavier@emulex.com>

in view of IP based GID changes, now vlan-id
can be obtained from qp_attr->vlan_id.

Signed-off-by: Devesh Sharma <devesh.sharma@emulex.com>
Signed-off-by: Selvin Xavier <selvin.xavier@emulex.com>
(cherry picked from commit 60bea5be1170fbaef27c6e07b6553db1eb15cb8b)
Signed-off-by: Doug Ledford <dledford@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/ocrdma/ocrdma_hw.c b/drivers/infiniband/hw/ocrdma/ocrdma_hw.c
index dd35ae5..10a261c 100644
--- a/drivers/infiniband/hw/ocrdma/ocrdma_hw.c
+++ b/drivers/infiniband/hw/ocrdma/ocrdma_hw.c
@@ -2279,7 +2279,8 @@ mbx_err:
 
 static int ocrdma_set_av_params(struct ocrdma_qp *qp,
     struct ocrdma_modify_qp *cmd,
-    struct ib_qp_attr *attrs)
+    struct ib_qp_attr *attrs,
+    int attr_mask)
 {
  int status;
  struct ib_ah_attr *ah_attr = &attrs->ah_attr;
@@ -2319,8 +2320,8 @@ static int ocrdma_set_av_params(struct ocrdma_qp *qp,
  ocrdma_cpu_to_le32(&cmd->params.dgid[0], sizeof(cmd->params.dgid));
  ocrdma_cpu_to_le32(&cmd->params.sgid[0], sizeof(cmd->params.sgid));
  cmd->params.vlan_dmac_b4_to_b5 = mac_addr[4] | (mac_addr[5] << 8);
- vlan_id = ah_attr->vlan_id;
- if (vlan_id && (vlan_id < 0x1000)) {
+ if (attr_mask & IB_QP_VID) {
+  vlan_id = attrs->vlan_id;
   cmd->params.vlan_dmac_b4_to_b5 |=
       vlan_id << OCRDMA_QP_PARAMS_VLAN_SHIFT;
   cmd->flags |= OCRDMA_QP_PARA_VLAN_EN_VALID;
@@ -2347,7 +2348,7 @@ static int ocrdma_set_qp_params(struct ocrdma_qp *qp,
   cmd->flags |= OCRDMA_QP_PARA_QKEY_VALID;
  }
  if (attr_mask & IB_QP_AV) {
-  status = ocrdma_set_av_params(qp, cmd, attrs);
+  status = ocrdma_set_av_params(qp, cmd, attrs, attr_mask);
   if (status)
    return status;
  } else if (qp->qp_type == IB_QPT_GSI || qp->qp_type == IB_QPT_UD) {
-- 
1.7.1